include/type_traits/traits_tpl.hpp

```diff
@@ -0,0 +1,267 @@
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Francesco Rizzi (fnrizzi@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef TPL_TRAITS_HPP_
+#define TPL_TRAITS_HPP_
+
+namespace pressio { namespace impl {
+
+#ifdef PRESSIO_ENABLE_TPL_TRILINOS
+
+//*******************************
+// Common traits
+//*******************************
+
+template<
+    int Rank,
+    typename Scalar,
+    typename LocalOrdinal,
+    typename GlobalOrdinal,
+    typename DataMap,
+    typename Comm
+>
+struct TrilinosTraits
+  : public ::pressio::impl::ContainersSharedTraits<
+      PackageIdentifier::Trilinos,
+      false,
+      Rank
+    >,
+    public ::pressio::impl::ScalarTrait<Scalar>,
+    public ::pressio::impl::DynamicAllocTrait
+{
+  using local_ordinal_type = LocalOrdinal;
+  using global_ordinal_type = GlobalOrdinal;
+  using data_map_type = DataMap;
+  using size_type  = GlobalOrdinal;
+  using communicator_type = Comm;
+};
+
+//*******************************
+// Epetra traits
+//*******************************
+template<int Rank = 1>
+using EpetraTraits = TrilinosTraits<
+  Rank,
+  double, int, int,
+  Epetra_BlockMap,
+  Epetra_Comm
+>;
+
+//*******************************
+// Tpetra traits
+//*******************************
+template<typename T>
+struct TrilinosCommType
+{
+  using type = decltype(
+    std::declval<
+      typename T::map_type
+    >().getComm()
+  );
+};
+
+template<typename T, typename enabled = void> struct TpetraExtraTraits {};
+
+template<typename T>
+struct TpetraExtraTraits<
+  T,
+  mpl::enable_if_t< // NOT Tpetra::BlockVector
+    is_vector_tpetra<T>::value or
+    is_multi_vector_tpetra<T>::value
+  >
+>
+{
+  using dual_view_type = typename T::dual_view_type;
+  using dot_type = typename T::dot_type;
+  using mag_type = typename T::mag_type;
+};
+
+template<typename T, int Rank = 1>
```

Comment:
can you remove default = 1? I like to force specifying it all the time.

Reply:
Fixed